src/Dfe.Spi.UkrlpAdapter.Infrastructure.UkrlpSoapApi/UkrlpSoapApiClient.cs

```diff
@@ -0,0 +1,80 @@
+﻿using System;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Xml.Linq;
+using Dfe.Spi.UkrlpAdapter.Domain.Configuration;
+using Dfe.Spi.UkrlpAdapter.Domain.UkrlpApi;
+using RestSharp;
+
+namespace Dfe.Spi.UkrlpAdapter.Infrastructure.UkrlpSoapApi
+{
+    public class UkrlpSoapApiClient : IUkrlpApiClient
+    {
+        private readonly IRestClient _restClient;
+        private readonly IUkrlpSoapMessageBuilder _messageBuilder;
+
+        internal UkrlpSoapApiClient(IRestClient restClient, IUkrlpSoapMessageBuilder messageBuilder,
+            UkrlpApiConfiguration configuration)
+        {
+            _restClient = restClient;
+            _restClient.BaseUrl = new Uri(configuration.Url, UriKind.Absolute);
+            _messageBuilder = messageBuilder;
+        }
+
+        public UkrlpSoapApiClient(IRestClient restClient, UkrlpApiConfiguration configuration)
+            : this(restClient, new UkrlpSoapMessageBuilder(configuration.StakeholderId), configuration)
+        {
+        }
+
+        public async Task<Provider> GetProviderAsync(long ukprn, CancellationToken cancellationToken)
+        {
+            var message = _messageBuilder.BuildMessageToGetSpecificUkprn(ukprn);
+
+            var request = new RestRequest(Method.POST);
+            request.AddParameter("text/xml", message, ParameterType.RequestBody);
+            request.AddHeader("SOAPAction", "retrieveAllProviders");
+
+            var response = await _restClient.ExecuteTaskAsync(request, cancellationToken);
+            var result = EnsureSuccessResponseAndExtractResult(response);
+
+            var match = result.GetElementByLocalName("MatchingProviderRecords");
+            if (match == null)
+            {
+                return null;
+            }
+
+            return new Provider
+            {
+                UnitedKingdomProviderReferenceNumber = ukprn,
+                ProviderName = match.GetElementByLocalName("ProviderName").Value,
+            };
+        }
+
+        private static XElement EnsureSuccessResponseAndExtractResult(IRestResponse response)
+        {
+            XDocument document;
+            try
+            {
+                document = XDocument.Parse(response.Content);
+            }
+            catch (Exception ex)
+            {
+                throw new UkrlpSoapApiException($"Error deserializing SOAP response: {ex.Message} (response: {response.Content})", ex);
+            }
+            
+            var envelope = document.Elements().Single();
+            var body = envelope.GetElementByLocalName("Body");
+
+            if (!response.IsSuccessful)
+            {
+                var fault = body.Elements().Single();
+                var faultCode = fault.GetElementByLocalName("faultcode");
+                var faultString = fault.GetElementByLocalName("faultstring");
```

Comment:
I imagine there's no .NET Core constants for these SOAP-specific things? I've just done a quick Google search, and it didn't throw up anything. Maybe there isn't!

Reply:
Not that I am aware of. Might be somewhere in WCF libraries

Reply:
Maybe, not really worth including a whole lib for some constants though, it's not like these values are going to change!